client/api/index.js

```diff
@@ -239,11 +239,86 @@ export default class WCStripeAPI {
 	 *
 	 * @param {string} redirectUrl The redirect URL, returned from the server.
 	 * @param {string} paymentMethodToSave The ID of a Payment Method if it should be saved (optional).
-	 * @return {string|boolean} A redirect URL on success, or `true` if no confirmation is needed.
+	 * @return {mixed} A redirect URL on success, or `true` if no confirmation is needed.
 	 */
 	confirmIntent( redirectUrl, paymentMethodToSave ) {
-		console.error( 'TODO: Not implemented yet: confirmIntent' );
-		return true;
+		const partials = redirectUrl.match(
```

Comment:
I don't think this would be possible in this case. We set the `order_id` via `wp_localize_script` to use with `createIntent` [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/api/index.js#L322-L351) on the Order Pay page, because the Order already exists before the page is loaded. This `confirmIntent` function is actually implemented (mostly) the [same way it is here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/api/index.js#L170-L269) and has a particularly unique purpose that took me a long time to realise needed to exist. 

This function is only called during the 3DS authentication flow with Stripe and I believe it will only be triggered during a checkout with a saved payment method. There are certain [cards that require additional authentication](https://stripe.com/docs/testing#regulatory-cards) at checkout before we can proceed with a transaction. Normally, this authentication is triggered by calling either `confirmPayment` or `confirmSetup` via the Stripe API, which happens at the [end of a regular UPE checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/client/classic/upe/index.js#L547-L551). 

![Authentication modal](https://cdn-std.droplr.net/files/acc_1104206/79pTga)
_Authentication modal appears when using 3DS test card_

However, if we save one of these cards as a payment token, we now no longer use the UPE or Stripe's JS to checkout and trigger this modal, but we still need to show the modal and authenticate the user before we can complete this transaction. The way that this is handled here--which is borrowed from WCPay's regular checkout flow--is that after confirming the intent, if the card requires further authentication, the status of the intent will be `requires_action`. This gets caught by [this block](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L540-L560) and we return [this URL parameter](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L551-L557) as a redirect URL for the response of `process_payment`. This means that the Woo checkout doesn't redirect to the Order Complete page, but instead just adds this parameter to the URL of the current page.

That URL change is caught by this [on `hashchange` event](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/client/classic/upe/index.js#L687-L691) and this fires the `maybeShowAuthenticationModal` function, which finally fires `confirmIntent`. _Now_ we can at last scrape the URL parameters to get the order ID and client secret used on the payment intent, which is what is happening here in `redirectUrl.match`. We use the client secret to fire [`confirmCardSetup` or `confirmCardPayment`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/client/api/index.js#L278-L280) via Stripe's JS API, which triggers the authentication modal at last. We use the response of this request to [submit a new AJAX request](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/client/api/index.js#L294) to [update the order status](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/includes/class-wc-stripe-intent-controller.php#L524-L588) and, finally, depending on the result of that AJAX request, we either [manually redirect the user to the Order Complete page](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/client/classic/upe/index.js#L600) or [we show an error on the checkout form](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2a8e109f6ab9aa6a04f4c87874ac458744ee2172/client/classic/upe/index.js#L615).

PHEW.

Apologies for another of my famous, long-winded explanations, but hopefully that explains why these weird partials here are necessary and hopefully I didn't just bore you with an explanation that you were already aware of. Let me know if this long ramble satisfies you or if you still think there is a better way to handle this case. 

P.S. This is why a left [this comment on this PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1780/files#r700723607) stating that I think the `paymentMethodToSave` parameter on the `confirmIntent` function is redundant, because this function will only be called when checking out with a payment method that has already been saved; however, I wasn't entirely sure that this function might not be called from elsewhere, so I left it in.